### PR TITLE
Version 3.2

### PR DESCRIPTION
**Pull request by krateng**
Main Features:

* Container uses linuxserver.io base
* Support for albums
* Asynchronous image fetching

Run this branch to test it out, I'm thankful for any feedback how well it works with real world data. If everything breaks, just don't [blame it on me](https://maloja.krateng.ch/track?artist=Twice&title=Blame+it+on+me) :)

The most important feature is album support. Right now, maloja accepts album information from new scrobbles; and you can try to parse old scrobbles with the command `maloja parsealbums` - this data will be available e.g. if you've imported from lastfm fairly recently. You can also do the usual web interface stuff like merging and renaming. Remember that css is often aggressively cached, so refresh that if the web pages look weird.

Things still to do

- [ ] Settings for picking album only by album title when no albumartist info is present
- [x] Way to quickly find album-less tracks for easier manual fixing
- [x] Album inclusion in startpage
- [ ] Implementation of majority album assignment instead of trust first or trust last

### CODE REVIEW

**Comment by FoxxMD** [comment]
Continuing the discussion from https://github.com/krateng/maloja/issues/96#issuecomment-1490562761

I think as long as there is a way to _unmerge_ track/artist combos from an album in the ui then whatever default behavior is chosen for merging albums wouldn't be an issue since it could always be fixed by the user.

Another option could be adding some additional `rule` types:

* `mergealbum` -> merge any album with the given value into any existing album with the same name
* `unmergealbum` -> current behavior -- create a new album if it has a unique artist combo from any found in an album with the same name

The values for these could take a special wildcard `*` to define default behavior

Always merge into an existing album
```tsv
mergealbum	*
```

Always create new albums for unique artist combos
```tsv
unmergealbum	*
```

**Comment by jan666** [comment]
Hi,

I just wanted to try it out but it looks like parsing old albums does not work for me, just for new ones.

```
$ maloja parsealbums
[sqldb] Column album_id was added to table tracks!
Parsing album information...
Adding 0 tracks to albums...
Done!
```

```
sqlite> select * from scrobbles order by timestamp desc limit 10;
1680240956|{"track_artists": ["The Offenders"], "track_title": "Carla e Paolo", "album_title": "Orthodoxy Of New Radicalism", "scrobble_time": 1680240956}|client:dbscrobbler||27675|{"album_title": "Orthodoxy Of New Radicalism", "album_artists": ["The Offenders"]}
1680240759|{"track_artists": ["The Offenders"], "track_title": "Hall of Fame", "album_title": "Orthodoxy Of New Radicalism", "scrobble_time": 1680240759}|client:dbscrobbler||27809|{"album_title": "Orthodoxy Of New Radicalism", "album_artists": ["The Offenders"]}
1680240581|{"track_artists": ["The Offenders"], "track_title": "Invisible", "album_title": "Orthodoxy Of New Radicalism", "scrobble_time": 1680240581}|client:dbscrobbler||27810|{"album_title": "Orthodoxy Of New Radicalism", "album_artists": ["The Offenders"]}
1680240413|{"track_artists": ["The Offenders"], "track_title": "Cardboard", "album_title": "Orthodoxy Of New Radicalism", "scrobble_time": 1680240413}|client:dbscrobbler||27811|{"album_title": "Orthodoxy Of New Radicalism", "album_artists": ["The Offenders"]}
1680239921|{"track_artists": ["The Offenders"], "track_title": "The Messenger", "album_title": "Orthodoxy Of New Radicalism", "scrobble_time": 1680239921}|client:dbscrobbler||27387|{"album_title": "Orthodoxy Of New Radicalism", "album_artists": ["The Offenders"]}
1680239732|{"track_artists": ["The Offenders"], "track_title": "You vs Reality", "album_title": "Orthodoxy Of New Radicalism", "scrobble_time": 1680239732}|client:dbscrobbler||27812|{"album_title": "Orthodoxy Of New Radicalism", "album_artists": ["The Offenders"]}
1680239585|{"track_artists": ["The Offenders"], "track_title": "Letter from the Front", "album_title": "Orthodoxy Of New Radicalism", "scrobble_time": 1680239585}|client:dbscrobbler||27813|{"album_title": "Orthodoxy Of New Radicalism", "album_artists": ["The Offenders"]}
1680239263|{"track_artists": ["The Offenders"], "track_title": "Tales From My Neighborhood", "album_title": "Orthodoxy Of New Radicalism", "scrobble_time": 1680239263}|client:dbscrobbler||27795|{"album_title": "Orthodoxy Of New Radicalism", "album_artists": ["The Offenders"]}
1680184991|{"track_artists": ["G\u00fcnther and the Jauchs"], "track_title": "Bunte Ponys", "album_name": "Menschheit, meh", "scrobble_time": 1680184991}|client:dbscrobbler||27758|
1680179956|{"track_artists": ["G\u00fcnther and the Jauchs"], "track_title": "Primogenitur Made Me Do It", "album_name": "Menschheit, meh", "scrobble_time": 1680179956}|client:dbscrobbler||27759|

```

**Comment by krateng** [comment]
Would you mind sharing your sql database file?

**Comment by jan666** [comment]
> Would you mind sharing your sql database file?

I just sent you an email

**Comment by krateng** [comment]
Thank you! Found the problem, will fix it asap.

**Comment by jan666** [comment]
```
[...]
13500 of 13534
Done!
```

👍

**Comment by darknesswithsorrowtears** [comment]
Can you add that when you upload an album cover, it will change for all the tracks in the album as well?

**Comment by krateng** [comment]
This should now be possible with the setting `USE_ALBUM_ARTWORK_FOR_TRACKS` (`Use Album Artwork for tracks` in the web interface)

**Comment by jan666** [comment]
I just built the latest version from `next_minor_version` and now `/image` does not work anymore:

```
Request URL: https://[...]/image?artist_id=1562
Request Method: GET
Status Code: 503 Service Unavailable
```

switched back to the branch `feature-albums` and it works again.

**Comment by krateng** [comment]
The 504 response should only be temporary while the image is being fetched. It should return the correct image eventually.

**Comment by darknesswithsorrowtears** [comment]
I think it's worth separating same-track scrobbles based on "album field". For at the moment, if I have a song in the album scrobbled, then while listening to this track in single or a compilation album, the scrobble will still go to the album, ignoring the fact that album field is different. So, more accurate scrobbling stats for Demos, Singles, Deluxe Albums, Compilations, etc. (last.fm do this)

**Comment by krateng** [comment]
The album info is designed to be non-defining, since it shouldn't matter in what context you listened to a song - if you've listened to it 50 times on the album, 50 times on the single, 50 times on the EP and 50 times on the Best Of compilation, it should still be ahead of a track you've only listened to 150 times. I am however planning to allow later album info to overwrite earlier scrobbles.

**Comment by duckfromdiscord** [comment]
`maloja parsealbums` is not working for me, I ran `pip install .` after cloning this `next_minor_version` branch. Seems to not be recognized. I installed it as `maloja_preview`:

```
~# maloja_preview
Valid commands: start restart stop run supervisor debug setup import backup generate export apidebug info
~# maloja_preview parsealbums
Valid commands: start restart stop run supervisor debug setup import backup generate export apidebug info
```

**Comment by duckfromdiscord** [comment]
Is it possible any commit in next_minor_version affected the listenbrainz api?

**Comment by duckfromdiscord** [comment]
There was some weird glitch with Japanese text (I think?) causing the scrobble list tables to become double-height and the 🗑️ 🔁 icons would be on top of each other vertically instead of next to each other horizontally. Did one of those CSS changes fix that? Because I can't tell if it's stopped happening or whatever song name was causing it just hasn't come up

**Comment by krateng** [comment]
I haven't fixed any CSS specifically for that, but with how chaotic my code is, everything could technically have all kinds of effects all over the application :D Let me know if it comes up again!

**Comment by duckfromdiscord** [comment]
Should I direct #241 to gonic or is this a maloja issue?

File: maloja/database/sqldb.py [hunk 1]
```diff
@@ -898,20 +910,30 @@ def get_scrobbles_of_album(album,since=None,to=None,resolve_references=True,dbco
...
-		DB['scrobbles'].c.timestamp>=since,
-	).order_by(sql.asc('timestamp'))
+		DB['scrobbles'].c.timestamp.between(since,to)
+	)
+	if reverse:
+		op = op.order_by(sql.desc('timestamp'))
+	else:
+		op = op.order_by(sql.asc('timestamp'))
+	if limit:
+		op = op.limit(limit)
```
Comment (FoxxMD) [inline maloja/database/sqldb.py:945]: I was about to suggest this, glad i checked changed first. Should help massively with performance for multi-scrobbler and, i'm assuming, "last scrobbles" load time for maloja

**Comment by duckfromdiscord** [comment]
To be sure, everything from master that had the new Spotify json format support is in this next minor version branch now? That means I can delete my import fix branch.

**Comment by jan666** [comment]
> The 503 response should only be temporary while the image is being fetched. It should return the correct image eventually.

I still have that issue with the `next_minor_version` branch. `feature_albums` branch works. Maloja don't even show the artist images I uploaded manually. 

I already tried deleting the cache folder.

This is from the log

```
[thirdparty] Could not get album image for (['BURNOUT OSTWEST'], 'Würzburg stärkt die Szene') from MusicBrainz
[thirdparty] Could not get album image for (['Goldzilla'], 'Goldzilla vs Dortmund') from MusicBrainz
[thirdparty] Could not get album image for (['Rancid'], 'Tomorrow Never Comes') from Deezer
[thirdparty] Could not get album image for (['Rancid'], 'Tomorrow Never Comes') from MusicBrainz
[thirdparty] Could not get album image for (['Kotzreiz'], 'Du machst die Stadt kaputt (Bonus Track Version)') from Deezer
[thirdparty] Could not get album image for (['Kotzreiz'], 'Du machst die Stadt kaputt (Bonus Track Version)') from MusicBrainz
[thirdparty] Could not get album image for (['Knochenfabrik'], 'Cooler Parkplatz') from Deezer
[thirdparty] Could not get album image for (['Knochenfabrik'], 'Cooler Parkplatz') from MusicBrainz
[thirdparty] Could not get album image for (['Kotzreiz'], 'Nüchtern unerträglich') from Deezer
[thirdparty] Could not get album image for (['Kotzreiz'], 'Nüchtern unerträglich') from MusicBrainz
[thirdparty] Could not get album image for (['Kotzreiz'], 'Punk bleibt Punk') from Deezer
[thirdparty] Could not get album image for (['Kotzreiz'], 'Punk bleibt Punk') from MusicBrainz
[thirdparty] Could not get album image for (['Pascow'], 'Diene der Party') from Deezer
[thirdparty] Could not get album image for (['Pascow'], 'Diene der Party') from MusicBrainz
[thirdparty] Could not get album image for (['Pascow'], 'Alles muss kaputt sein!') from Deezer
[thirdparty] Could not get album image for (['Pascow'], 'Alles muss kaputt sein!') from MusicBrainz
[thirdparty] Could not get album image for (['Pascow'], 'Nächster Halt gefliester Boden') from Deezer
[thirdparty] Could not get album image for (['Pascow'], 'Nächster Halt gefliester Boden') from MusicBrainz
[thirdparty] Could not get album image for (['Beatsteaks'], '23 Singles') from Deezer
[thirdparty] Could not get album image for (['Beatsteaks'], '23 Singles') from MusicBrainz
[thirdparty] Could not get album image for (['Pascow'], 'Jade') from Deezer
[thirdparty] Could not get album image for (['Pascow'], 'Jade') from MusicBrainz
[thirdparty] Could not get album image for (['Sondaschule'], 'Schön kaputt (Bonus Tracks Version)') from Deezer
[thirdparty] Could not get album image for (['Sondaschule'], 'Schön kaputt (Bonus Tracks Version)') from MusicBrainz
[thirdparty] Could not get album image for (['ZSK'], 'Herz für die Sache') from Deezer
[thirdparty] Could not get album image for (['ZSK'], 'Herz für die Sache') from MusicBrainz
[thirdparty] Could not get album image for (['Oxo 86'], 'Auf die Liebe und auf die Sehnsucht') from Deezer
[thirdparty] Could not get album image for (['Oxo 86'], 'Auf die Liebe und auf die Sehnsucht') from MusicBrainz
[thirdparty] Could not get album image for (['Die Liga der gewöhnlichen Gentlemen'], "Fuck Dance, Let's Art") from Deezer
[thirdparty] Could not get album image for (['Die Liga der gewöhnlichen Gentlemen'], "Fuck Dance, Let's Art") from MusicBrainz
[thirdparty] Could not get album image for (['ZSK'], 'Hallo Hoffnung') from Deezer
[thirdparty] Could not get album image for (['ZSK'], 'Hallo Hoffnung') from MusicBrainz
[thirdparty] Could not get album image for (['Los Fastidios'], 'Joy Joy Joy') from Deezer
[thirdparty] Could not get album image for (['Los Fastidios'], 'Joy Joy Joy') from MusicBrainz
[thirdparty] Got album image for (['Pascow'], 'Sieben') from Deezer
[thirdparty] Could not get album image for (['NTA'], 'Dark Stories - Pt. I - EP') from Deezer
[thirdparty] Got album image for (['KMPFSPRT'], 'Intervention') from Deezer
[thirdparty] Could not get album image for (['NTA'], 'Dark Stories - Pt. I - EP') from MusicBrainz
[thirdparty] Could not get album image for (['Nachkriegsratten'], 'Sandmann - Single') from Deezer
[thirdparty] Could not get album image for (['Nachkriegsratten'], 'Sandmann - Single') from MusicBrainz
[thirdparty] Could not get album image for (['Goldzilla'], 'Goldzilla vs die Blumen des Schreckens - EP') from Deezer
[thirdparty] Could not get album image for (['Goldzilla'], 'Goldzilla vs die Blumen des Schreckens - EP') from MusicBrainz
[thirdparty] Got album image for (['The Sensitives'], 'Punch') from Deezer
[thirdparty] Got album image for (['KMPFSPRT'], 'KMPFSPRT') from Deezer
[thirdparty] Got album image for (['Jack Pott'], 'Hass im Ärmel') from Deezer
```

**Comment by krateng** [comment]
I tried with the same albums at it works for me. Are you maybe running into the rate limit? It looks like you are getting some album images, so it seems to work.

**Comment by jan666** [comment]
There is no error in the log (running via `maloja debug`). No image is shown at all - why are my local artist images missing, too? 

Can I check if the images were downloaded at all?

**Comment by krateng** [comment]
Can you run the server in the foreground so you can see exceptions directly? Then just load album pages until it shows an error.

**Comment by jan666** [comment]
Ok this is really strange. I pulled the latest `next_minor_version`, installed, deleted `cache` folder, loaded Albums page:

```
[thirdparty] Could not get album image for (['Pershing Boys'], 'Pershing Boys (Demo)') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['100 Kilo Herz'], 'Weit weg von zu Hause') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Alarmsignal'], 'Sklaven der Langeweile') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Die Liga der gewöhnlichen Gentlemen'], 'Gschichterln aus dem Park Cafe') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Die Liga der gewöhnlichen Gentlemen'], "Fuck Dance, Let's Art") from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Ur-in-stinkt'], 'Ur-in-stinkt / Frost, Pisse & Elend - Split') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Disco//Oslo'], 'Disco//Oslo') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Nocte Obducta'], 'Karwoche - Die Sonne der Toten pulsiert') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Superpunk'], 'Why Not?') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Egotronic'], 'Stresz') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Einkauf Aktuell'], 'Blutender Mann vor Landschaft') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Malatesta'], 'Unter Tage') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Popperklopper'], 'Wahnsinn Weltweit') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Mantar'], 'Death by Burning') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['On the Huh'], 'Peepshow EP') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Punch'], "They Don't Have to Believe") from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Unhaim'], 'Chanchanchotzäbäbäbä... Stuss') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['SOKO LiNX'], 'Auf die Fresze. Fertig. Los!') from MusicBrainz
[thirdparty] Could not get album image for (['The Sensitives'], 'Punch') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Oxo 86'], 'Fröhlich sein und singen') from MusicBrainz
[thirdparty] Could not get album image for (['Los Fastidios'], 'The Sound of Revolution') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Scherben'], 'Domestiziert') from MusicBrainz
[thirdparty] Error getting album image from Deezer: Rate Limit exceeded
[thirdparty] Could not get album image for (['Fahnenflucht'], 'Angst und Empathie') from MusicBrainz
```

So it looks like rate limit. I stopped maloja, deleted `caches` folder again, checkout out and installed the `feature-albums` branch, loaded Albums page and it can download withput problems:

```
[server] Starting up Maloja server...
[server] Listening on *:42010
[debug_performance] Executed jinja_page (('charts_albums',), {}) in 0.86s
[thirdparty] Got album image for (['Pascow'], 'Sieben') from Deezer
[thirdparty] Got album image for (['Team Scheisse'], '042124192799') from Deezer
[thirdparty] Got album image for (['Chefdenker'], 'Asozialdarwinismus') from Deezer
[thirdparty] Got album image for (['Knochenfabrik'], 'Ameisenstaat') from Deezer
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/5ccacf656cbe3f7a8940d184fa58cb37/250x250-000000-80-0-0.jpg for local caching
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/ea6bf9342fafdb9bc127c7a5720608f6/250x250-000000-80-0-0.jpg for local caching
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/4484fb2a3bf61b85f0e033f4628a9016/250x250-000000-80-0-0.jpg for local caching
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/160294918b0236632a57da3206dbecb3/250x250-000000-80-0-0.jpg for local caching
[thirdparty] Got album image for (['Rantanplan'], 'Ahoi') from Deezer
[thirdparty] Got album image for (['duesenjaeger'], 'Die Gespenster und der Schnee') from Deezer
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/cd7fb4ce4ae8cb1747f86a595124ac6b/250x250-000000-80-0-0.jpg for local caching
[thirdparty] Got album image for (['BURNOUT OSTWEST'], 'Würzburg stärkt die Szene') from Deezer
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/aa724b2dabfd965ee2d515aa18b2e82f/250x250-000000-80-0-0.jpg for local caching
[thirdparty] Got album image for (['BSK'], 'PUNK in dieser Form') from Deezer
[thirdparty] Got album image for (['Goldzilla'], 'Goldzilla vs Dortmund') from Deezer
[thirdparty] Got album image for (['Rancid'], 'Tomorrow Never Comes') from Deezer
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/ed30b6da1f63358bced584a5f9a74910/250x250-000000-80-0-0.jpg for local caching
[thirdparty] Could not get album image for (['Kotzreiz'], 'Du machst die Stadt kaputt (Bonus Track Version)') from Deezer
[thirdparty] Could not get album image for (['Kotzreiz'], 'Du machst die Stadt kaputt (Bonus Track Version)') from MusicBrainz
[thirdparty] Got album image for (['Kotzreiz'], 'Nüchtern unerträglich') from Deezer
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/8ca0211b466dd763537ea06c394aa822/250x250-000000-80-0-0.jpg for local caching
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/55bb1b74e53b3977e186520d795547f5/250x250-000000-80-0-0.jpg for local caching
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/0018a55235b56f6523e9fdb2bc51a0cd/250x250-000000-80-0-0.jpg for local caching
[thirdparty] Got album image for (['Kotzreiz'], 'Punk bleibt Punk') from Deezer
[thirdparty] Got album image for (['Knochenfabrik'], 'Cooler Parkplatz') from Deezer
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/ba3c7d2e3465ac3ae4bfb62ada6ab5c4/250x250-000000-80-0-0.jpg for local caching
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/33d02764e624a7909014ce1500ed493a/250x250-000000-80-0-0.jpg for local caching
[thirdparty] Could not get album image for (['Günther and the Jauchs'], 'Menschheit, meh') from Deezer
[thirdparty] Could not get album image for (['Günther and the Jauchs'], 'Menschheit, meh') from MusicBrainz
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/516308310e657104ef6ec336d75148cb/250x250-000000-80-0-0.jpg for local caching
[thirdparty] Could not get album image for (['Pershing Boys'], 'Zuviel Hoffnung') from Deezer
[thirdparty] Could not get album image for (['Pershing Boys'], 'Zuviel Hoffnung') from MusicBrainz
[thirdparty] Got album image for (['magdebored'], 'Es bleibt: komplex') from Deezer
[sqldb] Database Cleanup...
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/03187695c5c40f7067dc6463dae7a562/250x250-000000-80-0-0.jpg for local caching
[thirdparty] Got album image for (['KMPFSPRT'], 'Euphorie und Panik') from Deezer
[thirdparty] Got album image for (['Pascow'], 'Nächster Halt gefliester Boden') from Deezer
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/677b4ebefc46f32b2637d59a586ce3c6/250x250-000000-80-0-0.jpg for local caching
[images] Downloaded https://e-cdns-images.dzcdn.net/images/cover/674d48de2d075727980dd3b887c8eb44/250x250-000000-80-0-0.jpg for local caching
```

For me that does not really make sense?

**Comment by krateng** [comment]
The new version is much faster at actually getting images, but hitting the Deezer rate limit seems pretty crazy. Guess I'll add some delay or something.

**Comment by jan666** [comment]
But I get just zero images on the new version - the tables stay empty. And my webuploads are gone, too. And should the rate limit not affect the old version too when I ran into it? (I guess its an ip address rate limit)

**Comment by krateng** [comment]
The rate limit is super short-term, just blocks you for a few seconds. Is the log above from your console or the logfile? If you don't have these tables in your database, there has to be some exception thrown somewhere.

**Comment by jan666** [comment]
I just deleted cache again, switched to updated `next_mintor_version` and installed it, started maloja in the foreground and just loaded a single album:

.../album?artist=Pascow&albumtitle=Sieben

the log says:

```
$ cat thirdparty.log
2023/10/25 11:26:00  Got album image for (['Pascow'], 'Sieben') from Deezer
```

but I cant load the image:

`Failed to load resource: the server responded with a status of 503 (Service Unavailable)`

The tables are there but they are empty

```
$ sqlite3 cache/imagecache.sqlite
SQLite version 3.43.1 2023-09-11 12:01:27
Enter ".help" for usage hints.
sqlite> .tables
albums   artists  tracks
sqlite> select count(*) from albums;
0
sqlite> select count(*) from artists;
0
sqlite> select count(*) from tracks;
0
sqlite>
```

I can't see any error messages.

**Comment by krateng** [comment]
So just to make sure, when you start maloja having deleted the cache db first, it does actually create the file though? And when you load an image, the browser dev tools say it returns 503 but the terminal doesn't show any error?

**Comment by jan666** [comment]
I just did a fresh start.

Created an empty folder, set `MALOJA_DATA_DIRECTORY` to this new crated folder.

Started maloja and did a single scrobble via webinterface for Pascow - Himmelhunde from album Sieben.

```
$ maloja run

Maloja v3.1.5
https://github.com/krateng/maloja

Various external services can be used to display images. If not enough of them are set up, only local images will be used.
        Please enter your Last.fm API Key. If you do not want to use one at this moment, simply leave this empty and press Enter.

        Please enter your Spotify API ID. If you do not want to use one at this moment, simply leave this empty and press Enter.

        Please enter your Spotify API Secret. If you do not want to use one at this moment, simply leave this empty and press Enter.

        Please enter your TheAudioDB API Key. If you do not want to use one at this moment, simply leave this empty and press Enter.

Do you want to set up a key to enable scrobbling? Your scrobble extension needs that key so that only you can scrobble tracks to your database. [Y/n]
n
Password has been set.
[server] Starting up Maloja server...
[server] Listening on *:42010
[debug_performance] Executed jinja_page (('start',), {}) in 0.22s
[debug_performance] Executed jinja_page (('admin_overview',), {}) in 0.01s
[debug_performance] Executed jinja_page (('admin_manual',), {}) in 0.00s
[database] Incoming scrobble [Client: browser | API: native/v1]: {'track_artists': ['Pascow'], 'track_title': 'Himmelhunde', 'album_title': 'Sieben'}
[dbcache] Invalidated 81 of 81 DB cache entries
[dbcache] Invalidated 1 of 1 DB cache entries
[debug_performance] Executed jinja_page (('start',), {}) in 0.27s
[thirdparty] Got album image for (['Pascow'], 'Sieben') from Deezer
[thirdparty] Got artist image for Pascow from Deezer
```
but the browser does not load the image:

`Failed to load resource: the server responded with a status of 503 (Service Unavailable)`

The cache database file is created - but all tables are empty:

```
test-maloja-data $ sqlite3 cache/imagecache.sqlite
SQLite version 3.43.1 2023-09-11 12:01:27
Enter ".help" for usage hints.
sqlite> .tables
albums   artists  tracks
sqlite> select count(*) from albums;
0
sqlite> select count(*) from artists;
0
sqlite> select count(*) from tracks;
0
```

**Comment by krateng** [comment]
Now that is strange. Are you running this on the same machine that you did the database inspection above? This should be available with all versions from 3.35.

**Comment by krateng** [comment]
Can you try updating sqlalchemy to something 2.*?

**Comment by jan666** [comment]
Yes. That fixed it... maybe it's time to update requirements.txt?

**Comment by jan666** [comment]
I just noticed its impossible to `import` a just exported export to a new instance of maloja:

```
stuff@python:~/maloja $ maloja export
Exported 117558 Scrobbles to /usr/home/stuff/maloja/maloja_export_2023_10_27_08_45_49.json
stuff@python:~/maloja $ mkdir test-maloja-data
stuff@python:~/maloja $ export MALOJA_DATA_DIRECTORY=/home/stuff/maloja/test-maloja-data/
stuff@python:~/maloja $ maloja setup

Maloja v3.1.5
https://github.com/krateng/maloja

Various external services can be used to display images. If not enough of them are set up, only local images will be used.
        Please enter your Last.fm API Key. If you do not want to use one at this moment, simply leave this empty and press Enter.

        Please enter your Spotify API ID. If you do not want to use one at this moment, simply leave this empty and press Enter.

        Please enter your Spotify API Secret. If you do not want to use one at this moment, simply leave this empty and press Enter.

        Please enter your TheAudioDB API Key. If you do not want to use one at this moment, simply leave this empty and press Enter.

Do you want to set up a key to enable scrobbling? Your scrobble extension needs that key so that only you can scrobble tracks to your database. [Y/n]
n
Password has been set.
Setup complete!
stuff@python:~/maloja $ maloja import /usr/home/stuff/maloja/maloja_export_2023_10_27_08_45_49.json
File /usr/home/stuff/maloja/maloja_export_2023_10_27_08_45_49.json could not be identified as a valid import source.
```

**Comment by krateng** [comment]
Should be fixed!

**Pull request merged**